### PR TITLE
Extension port

### PR DESCRIPTION
**Pull request by karelhala**
## Description
This will make hac-dev as plugin in hac-core. There is no longer AppEntry and we do not wrap components in providers since all of that should be done in hac-core.

## Type of change

- [x] Feature

## How to test or reproduce?
It should be enough just to install latest deps and run `ENVIRONMENT=prod npm run start:beta` (we have one last PR to fully support this https://github.com/openshift/hac-core/pull/35)

## Known limitations
There are currently two know limitations:
- no dry run for create resource (@vojtechszocs  @andrewballantyne are there any plans in supporting this?)
- no redux, we are working on adding redux to hac-core and will keep you updated on the process

### CODE REVIEW

**Comment by karelhala** [comment]
@andrewballantyne oooh, will do that!

File: config/dev.webpack.config.js
Comment (andrewballantyne) [inline config/dev.webpack.config.js:60]: I don't think this change is intended...

File: config/prod.webpack.config.js
Comment (andrewballantyne) [inline config/prod.webpack.config.js:8]: Think we want the other prefix here too.

File: package.json
Comment (andrewballantyne) [inline package.json:18]: Don't think we want to remove scripts they added @karelhala ...

File: src/Routes.tsx
Comment (andrewballantyne) [inline src/Routes.tsx:53]: I _think_ you're missing `/create` and `/applicatons` routes in the extension definition?

File: package.json
Comment (andrewballantyne) [inline package.json:53]: @karelhala This seems a little aggressive to modify their redux usage. It's fair to remove it because they can't really use Redux because they don't own the context. But I suspect you just pulled a feature out from under them 🤔 

@rohitkrai03 what's the use-case for your redux usage? Can you guys pivot to a React Context? That's a supported extension type and should persist between routes.

File: src/hooks/useActiveNamespace.ts
Comment (andrewballantyne) [inline src/hooks/useActiveNamespace.ts:12]: There is `k8sListResourceItems`: https://github.com/openshift/dynamic-plugin-sdk/blob/main/packages/lib-utils/src/k8s/k8s-resource.ts#L152

No need to destruct it.

File: src/hooks/useApplicationsInfo.ts
Comment (andrewballantyne) [inline src/hooks/useApplicationsInfo.ts:21]: @rohitkrai03 Do you guys have K8sResourceCommon-esk types?

File: src/pages/TestK8s.tsx
Comment (andrewballantyne) [inline src/pages/TestK8s.tsx:72]: ConfigMaps don't exist anymore. We should bring over our Application changes if they want to keep this page.

File: src/utils/__tests__/create-utils.spec.ts
Comment (andrewballantyne) [inline src/utils/__tests__/create-utils.spec.ts:12]: Still mocking the old file...

File: src/utils/create-utils.ts
Comment (andrewballantyne) [inline src/utils/create-utils.ts:239]: Same point here about `k8sListResourceItems`

File: src/utils/delete-utils.ts
Comment (andrewballantyne) [inline src/utils/delete-utils.ts:21]: Commented code?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Top level review. Not done to any real detail or via testing the code.

File: config/dev.webpack.config.js
Comment (karelhala) [reply config/dev.webpack.config.js:60]: Well the thing is that console.dot is wrapping each app with its prefix. Since this is not directly an app, but extension it inherits the prefix of `hacCore` but I guess we could update it as `.hacCore, .hacDev`. Or (which might be better imo) wrapping each `route` with `Container` with the app's name in `hac-core`

File: package.json
Comment (karelhala) [reply package.json:18]: This is inherited from `starter-app` we are no longer doing these builds.

File: package.json
Comment (karelhala) [reply package.json:53]: I've looked trough the code and currently redux is used to show notification messages. Until we have full support of redux I don't think we can keep these in.

File: src/hooks/useActiveNamespace.ts
Comment (karelhala) [reply src/hooks/useActiveNamespace.ts:12]: Oooh! Okay, will do that!

File: src/utils/delete-utils.ts
Comment (karelhala) [reply src/utils/delete-utils.ts:21]: Aaah, yeah!

File: config/dev.webpack.config.js
Comment (karelhala) [reply config/dev.webpack.config.js:60]: Opened a PR in hac-core to allow passing className as wrapper of current route. This way one route styling won't affect styling of different route.

File: src/Routes.tsx
Comment (karelhala) [reply src/Routes.tsx:53]: Ah good point!

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: src/hooks/useApplicationsInfo.ts
Comment (andrewballantyne) [reply src/hooks/useApplicationsInfo.ts:21]: @rohitkrai03 You will get common types for sure... but use-case specific types is beyond the scope of the SDK. You'll need to base your types off a common one in your world for your own use-cases. If we supported all types than the SDK would need to be the source of truth for several projects/plugins and it becomes a bit unmanageable and violates the separation of concerns principle.

I do believe we have an issue with exporting the `K8sResourceCommon` type right now, but it should be there.

File: config/dev.webpack.config.js
Comment (karelhala) [reply config/dev.webpack.config.js:28]: Good point! I should remove the comments, this is usefull for local development.

File: config/plugins.js
Comment (karelhala) [reply config/plugins.js]: Not anymore! Removed.

File: src/components/AddComponent/utils.ts
Comment (karelhala) [reply src/components/AddComponent/utils.ts:37]: Unfortunately not. We either have to tap in the hac-core redux or write custom one.

**Comment by karelhala** [comment]
@rohitkrai03 @andrewballantyne I've started to use the newest webpack plugin and am looking if we could somehow bridge redux to your app so you can utilize the notifications utilities.

**Comment by rohitkrai03** [comment]
> @rohitkrai03 @andrewballantyne I've started to use the newest webpack plugin and am looking if we could somehow bridge redux to your app so you can utilise the notifications utilities.

@karelhala Based on UX inputs we would like to remove the notification utilities anyways. So its okay to replace them with `console.log` statements for now.

File: config/plugins.js
Comment (andrewballantyne) [inline config/plugins.js:2]: Why Async?

File: src/dynamic-plugin-sdk.ts
Comment (andrewballantyne) [inline src/dynamic-plugin-sdk.ts:9]: Should be able to drop this file entirely. I was able to delete this variant in HAC-Core when I pulled alpha3

File: src/pages/TestK8s.tsx
Comment (andrewballantyne) [inline src/pages/TestK8s.tsx:13]: These are available now.
```suggestion
import { K8sModelCommon, K8sResourceCommon } from '@openshift/dynamic-plugin-sdk-utils';
```

File: src/pages/TestK8s.tsx
Comment (andrewballantyne) [reply src/pages/TestK8s.tsx:72]: On second thought -- we can do this later if needed -- let us avoid this overhead for now.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: config/plugins.js
Comment (karelhala) [reply config/plugins.js:2]: The webpack config uses commonJS (we can move it to esmodule if we want to, but that is probably outside of scope of this PR) and this plugin is ES module so we have to use `import` as a fallback

File: src/dynamic-plugin-sdk.ts
Comment (andrewballantyne) [reply src/dynamic-plugin-sdk.ts:10]: We should identify those and if needed bring those over to the SDK... we shouldn't keep this file, it was very much a hack.

File: config/plugins.js
Comment (andrewballantyne) [reply config/plugins.js:2]: Ah right -- that's unfortunate. Okay, thanks for the info.

File: src/dynamic-plugin-sdk.ts
Comment (andrewballantyne) [reply src/dynamic-plugin-sdk.ts:10]: They shouldn't need `commonFetch` either -- if they do need it they are not properly hooked into HAC-Core.

File: src/models/application.ts
Comment (andrewballantyne) [inline src/models/application.ts:10]: Hmm, I wonder if these are needed fields @rohitkrai03 if the Plugin adds them. These are half-holdovers from OpenShift Console to use models to show details on pages. I'm not a huge fan of this model, but the alternative is probably some mixture map to values.

Do you guys have a need for ResourceIcons? What about pages that read models and show labels like a page header?

File: src/models/component.ts
Comment (andrewballantyne) [inline src/models/component.ts:10]: same here

File: src/models/component.ts
Comment (andrewballantyne) [inline src/models/component.ts:27]: and here

File: src/models/project.ts
Comment (andrewballantyne) [inline src/models/project.ts:9]: and here

File: src/models/service-provider-integration.ts
Comment (andrewballantyne) [inline src/models/service-provider-integration.ts:10]: and here

**Review by andrewballantyne — COMMENTED** [review]
Other uses to my previous comment

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Only one more to my previous comments I noticed

File: src/shared/components/catalog/utils/catalog-utils.tsx
Comment (karelhala) [reply src/shared/components/catalog/utils/catalog-utils.tsx:54]: Good point! I've commented it to see what was causing the test fails. Uncommented it!

File: config/remotePlugin.js
Comment (rohitkrai03) [inline config/remotePlugin.js:13]:
```suggestion
      AppEntry: resolve(__dirname, '../src/components/AppFlow.tsx'),
```

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
Tried running the PR locally. For some reason the requests to fetch applications is failing on the application list page.
Screenshot 2022-03-10 at 9 05 08 PM

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@karelhala A few more issues I noticed.
- I still see the issues with CSS and all the routes seem to be rendered inside a grey box (padding on all sides). 
- We have the `AppBanner` that says `Development Preview` which is supposed to be rendered on all app studio pages. This is gone now because we do not have a `App.tsx` anymore. Is there a way we can do that now in this plugin model?

**Comment by karelhala** [comment]
@rohitkrai03 for the grey background, that's platform specific and should be present on all pages your UX should discuss this with wider audience, if you want I can give you point of contacts. As for the Dev preview, that shouldn't be too hard to add. The CSS might be a bit more tricky.

**Comment by rohitkrai03** [comment]
> @rohitkrai03 for the grey background, that's platform specific and should be present on all pages your UX should discuss this with wider audience, if you want I can give you point of contacts. As for the Dev preview, that shouldn't be too hard to add. The CSS might be a bit more tricky.

I am okay with the overall grey background, but the issue is whatever components we are trying to render it gets rendered within some padding on all sides. That breaks the UX that we have, I think we should be able to use the full width of the content section without any extra padding on all sides.

For comparison take a look at the `console.redhat.com` page and the hac-dev page below - 

HAC Dev - 
Screenshot 2022-03-11 at 8 54 34 PM

Console.redhat.com (Notice the title section on top is rendered full width and doesn't have a padding.)
Screenshot 2022-03-11 at 8 54 44 PM

**Comment by karelhala** [comment]
@rohitkrai03 I have figured out why the classes were not properly applied. Turns out the problem was previously some styling was applied in AppEntry, which is no longer used.

@andrewballantyne I've also updated hac-core to wrap the dynamic component with `section` instead of `Main` so we keep the styling same as before https://github.com/openshift/hac-core/pull/44

**Comment by karelhala** [comment]
@rohitkrai03 updated the code, should be running now. The custom styles will be applied once the hac-core PR is merged

File: package.json
Comment (jrichter1) [inline package.json:12]: We've added these parameters specifically to avoid the tests hanging on openshift-ci. Removing them might reintroduce the issue.

Comment (karelhala) [reply package.json:12]: I can try adding them back, but when I try to run it locally with them it hangs on the very first test.

**Comment by karelhala** [comment]
@rohitkrai03 The `codecov/patch` report showing 45% of diff is strange. I had to re-arrange a lot of code by using different imports. Looks like these lines were never tested and now it fails the task. Should I go hunting these lines and write tests for them?

**Comment by rohitkrai03** [comment]
> @rohitkrai03 The `codecov/patch` report showing 45% of diff is strange. I had to re-arrange a lot of code by using different imports. Looks like these lines were never tested and now it fails the task. Should I go hunting these lines and write tests for them?

@karelhala I think it's okay to manually merge this if codecov is becoming an issue.

**Comment by karelhala** [comment]
@rohitkrai03 mhh, yeah. There has been quite a few changes to routing, updated. Thank you for taking a look!

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@karelhala I found some more issues in the workflows - 
- Cannot view build logs for any of the components. Its always saying `No pipeline runs found` even though the pipeline run is available for that component.
- `Add Component` button doesn't take the user to create component form. It takes the user to `Create application` form.
- The `ComponentQueryDetection` doesn't work for any github URL.

https://user-images.githubusercontent.com/6041994/159274934-cab1fdb4-89dc-4ef4-b5a1-311bb8802f91.mov

**Comment by karelhala** [comment]
> Add Component button doesn't take the user to create component form. It takes the user to Create application form.

This is the same on master as well, it leads to `create?application=${application}` shouldn't this lead to `create?component=${application}`?

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]
@karelhala Overall looks good. A few remaining issues that I found -
- The view logs modal is now in the content centre instead of app centre.
- Component delete operation does not work anymore. 

Screenshot 2022-03-21 at 10 53 16 PM

https://user-images.githubusercontent.com/6041994/159330789-1f42408f-9501-473a-873e-616c6e6c83bb.mov

**Comment by karelhala** [comment]
@rohitkrai03 thank you! Fixed both of these issues, we don't have to append the modal to a component instead it's better to pass the class prefix to it so styles are applied. As for the delete operation for some reason the API version was not correct.

**Review by rohitkrai03 — APPROVED** [review]
Overall works fine. There are some CSS and layout issues that we'll follow up on another PR. @karelhala Thanks for the effort here for converting hac-dev to hac-core plugin.

/lgtm
/approve

**Pull request merged**
